File: WdtcCore/src/main/java/org/wdt/wdtc/core/download/fabric/FabricDonwloadInfo.java
package org.wdt.wdtc.core.download.fabric;


import com.google.gson.JsonObject;
import lombok.Getter;
import lombok.Setter;
import org.apache.log4j.Logger;
import org.wdt.utils.gson.JsonUtils;
import org.wdt.wdtc.core.download.infterface.InstallTaskInterface;
import org.wdt.wdtc.core.download.infterface.ModDownloadInfoInterface;
import org.wdt.wdtc.core.download.infterface.VersionJsonObjectInterface;
import org.wdt.wdtc.core.game.Launcher;
import org.wdt.wdtc.core.manger.DownloadSourceManger;
import org.wdt.wdtc.core.manger.FileManger;
import org.wdt.wdtc.core.utils.DownloadUtils;
import org.wdt.wdtc.core.utils.ModUtils;
import org.wdt.wdtc.core.utils.WdtcLogger;

import java.io.File;
import java.io.IOException;

public class FabricDonwloadInfo implements ModDownloadInfoInterface {
  private static final Logger logmaker = WdtcLogger.getLogger(FabricDonwloadInfo.class);
  @Getter
  protected final String FabricVersionNumber;
  protected final Launcher launcher;
  @Getter
  @Setter
  private FabricAPIDownloadTask ApiDownloadTask;


  public FabricDonwloadInfo(Launcher launcher, String FabricVersionNumber) {
    this.FabricVersionNumber = FabricVersionNumber;
    this.launcher = launcher;
  }

  public FabricDonwloadInfo(Launcher launcher, VersionJsonObjectInterface versionJsonObjectInterface) {
    this(launcher, versionJsonObjectInterface.getVersionNumber());
  }

  public String getFabricVersionFileUrl() {
    return DownloadSourceManger.getOfficialDownloadSource().getFabricMetaUrl() + "v2/versions/loader/%s/%s/profile/json";
  }


  public void DownloadProfileZip() {
    DownloadUtils.StartDownloadTask(getProfileZipUrl(), getProfileZipFile());
  }

  public String getProfileZipFile() {
    return String.format(FileManger.getWdtcCache() + "/%s-%s-frofile-zip.zip", launcher.getVersionNumber(), getFabricVersionNumber());
  }

  public String getProfileZipUrl() {
    return String.format(DownloadSourceManger.getOfficialDownloadSource().getFabricMetaUrl() + "v2/versions/loader/%s/%s/profile/zip", launcher.getVersionNumber(), getFabricVersionNumber());
  }

  public String FromFabricLoaderFolder() {
    return String.format("fabric-loader-%s-%s", getFabricVersionNumber(), launcher.getVersionNumber());
  }

  public File getFabricJar() {
    return new File(FileManger.getWdtcCache() + "/" + FromFabricLoaderFolder() + "/" + FromFabricLoaderFolder() + ".jar");
  }

  public File getFabricVersionJson() {
    return new File(String.format(FileManger.getWdtcCache() + "/%s-fabric-%s.json", launcher.getVersionNumber(), FabricVersionNumber));
  }

  public JsonObject getFabricVersionJsonObject() throws IOException {
    return JsonUtils.getJsonObject(getFabricVersionJson());
  }


  public boolean isAPIDownloadTaskNoNull() {
    return ApiDownloadTask != null;
  }

  @Override
  public String getModVersion() {
    return FabricVersionNumber;
  }

  @Override
  public InstallTaskInterface getModInstallTask() {
    return new FabricInstallTask(launcher, FabricVersionNumber);
  }

  @Override
  public ModUtils.KindOfMod getModKind() {
    return ModUtils.KindOfMod.FABRIC;
  }
}

File: WdtcCore/src/main/kotlin/org/wdt/wdtc/core/download/fabric/FabricDonwloadInfo.kt
package org.wdt.wdtc.core.download.fabric

import com.google.gson.annotations.JsonAdapter
import org.wdt.utils.gson.readFileToClass
import org.wdt.wdtc.core.download.infterface.ModDownloadInfoInterface
import org.wdt.wdtc.core.download.infterface.ModInstallTaskInterface
import org.wdt.wdtc.core.download.infterface.VersionsJsonObjectInterface
import org.wdt.wdtc.core.game.Version
import org.wdt.wdtc.core.game.serializeGameRuntimeDependencyGsonBuilder
import org.wdt.wdtc.core.manger.officialDownloadSource
import org.wdt.wdtc.core.manger.wdtcCache
import org.wdt.wdtc.core.utils.KindOfMod
import org.wdt.wdtc.core.utils.gson.DownloadInfoTypeAdapter
import org.wdt.wdtc.core.utils.noNull
import org.wdt.wdtc.core.utils.startDownloadTask
import org.wdt.wdtc.core.utils.toURL
import java.io.File
import java.net.URL

@JsonAdapter(DownloadInfoTypeAdapter::class)
open class FabricDonwloadInfo(
	protected val version: Version,
	override val modVersion: String
) : ModDownloadInfoInterface {
	var apiDownloadTask: FabricAPIDownloadTask? = null
	
	
	constructor(version: Version, versionsJsonObjectInterface: VersionsJsonObjectInterface) : this(
		version,
		versionsJsonObjectInterface.versionNumber.noNull()
	)
	
	val fabricVersionFileUrl: String
		get() = "${officialDownloadSource.fabricMetaUrl}v2/versions/loader/%s/%s/profile/json"
	
	suspend fun startDownloadProfileZip() {
		startDownloadTask(profileZipUrl, profileZipFile)
	}
	
	
	private val profileZipFile: File
		get() = File(wdtcCache, "${version.versionNumber}-$modVersion-frofile-zip.zip")
	private val profileZipUrl: URL
		get() = "${officialDownloadSource.fabricMetaUrl}v2/versions/loader/%s/%s/profile/zip".format(
			version.versionNumber,
			modVersion
		).toURL()
	
	
	private val fromFabricLoaderFolder: String
		get() = "fabric-loader-%s-%s".format(modVersion, version.versionNumber)
	
	
	val fabricJar: File
		get() = File(wdtcCache, "$fromFabricLoaderFolder/$fromFabricLoaderFolder.jar")
	val fabricVersionJson: File
		get() = File(wdtcCache, "${version.versionNumber}-fabric-$modVersion.json")
	
	val fabricProfileJsonObject: FabricProfileJsonObject
		get() = fabricVersionJson.readFileToClass(serializeGameRuntimeDependencyGsonBuilder)
	val isHasApiDownloadTask: Boolean
		get() = apiDownloadTask != null
	override val modInstallTask: ModInstallTaskInterface
		get() = FabricInstallTask(version, modVersion)
	override val modKind: KindOfMod
		get() = KindOfMod.FABRIC
	
	override fun toString(): String {
		return "FabricDonwloadInfo(modVersion='$modVersion', apiDownloadTask=$apiDownloadTask)"
	}
	
	override fun equals(other: Any?): Boolean {
		if (this === other) return true
		if (other !is FabricDonwloadInfo) return false
		
		if (modVersion != other.modVersion) return false
		if (apiDownloadTask != other.apiDownloadTask) return false
		
		return true
	}
	
	override fun hashCode(): Int {
		var result = modVersion.hashCode()
		result = 31 * result + (apiDownloadTask?.hashCode() ?: 0)
		return result
	}
	
	
}
